File: LC169_MajorityElement/majorityElement.cpp
/*
 * majorityElement.cpp
 *
 *  Created on: 2017年4月9日
 *      Author: BrianzChang
 */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int majorityElement(vector<int>& nums) {
    	int len = nums.size();
    	if (len == 1)
    		return nums[0];

    	int candidate = 0;
    	int cnt = 0;
    	for (int i=0; i<len; i++) {
    		if ((candidate != nums[i]) && (cnt == 0))
    			candidate = nums[i];
    		else if (candidate != nums[i])
    			cnt -= 1;
    		else if (candidate == nums[i])
    			cnt += 1;
    	}
    	return candidate;
    }
};

int main() {
	// Generate testing data
	int arr[] = {0, 0, 1, 2, 0, 3, 0, 4, 0, 0};
	vector<int> nums(arr, arr + sizeof(arr)/sizeof(arr[0]));

	// Process
	Solution mySolution;
	int answer = mySolution.majorityElement(nums);
	// Print
	for (vector<int>::iterator it=nums.begin(); it!=nums.end(); ++it){
		cout << *it << "\t";
	}
	cout << endl << answer << endl;

	return 0;
}

File: LC169_MajorityElement/majorityElement.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Mon Mar 27 21:22:58 2017

@author: BrianzChang
"""

class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        """ Moore voting algorithm"""
        
        # 82 ms
        candidate = 0
        cnt = 0
        for item in nums:
            if item != candidate and cnt == 0:
                candidate = item
                cnt = 1
            elif item != candidate:
                cnt -= 1
            elif item == candidate:
                cnt += 1
            else:
                print "There must be sth wrong"
        return candidate
        """
        # 88ms, 66ms, 52ms
        candidate, cnt = 0, 0
        for item in nums:
            if item == candidate:
                cnt += 1
            elif cnt == 0:
                candidate, cnt = item, 1
            else:
                cnt -= 1
        return candidate
        “”“

mySolution = Solution()
list_test = [0, 3, 4, 0, 0, 0, 0, 0, 3, 7, 8];
print mySolution.majorityElement(list_test)